src/ucp/wireup/wireup_cm.c

```diff
@@ -0,0 +1,272 @@
+        status = ucp_wireup_ep_create(ucp_ep, &ucp_ep->uct_eps[lane_idx]);
+        if (status != UCS_OK) {
+            break;
+        }
+    }
+
+    if (status != UCS_OK) {
+        for (lane_idx = 1; (lane_idx < ucp_ep_num_lanes(ucp_ep)) &&
+                           (ucp_ep->uct_eps[lane_idx] != NULL); ++lane_idx) {
+            uct_ep_destroy(ucp_ep->uct_eps[lane_idx]);
+        }
+    }
+
+free_addr_indicies:
+    ucs_free(addr_indices);
+free_addr_list:
+    ucs_free(unpacked_addr.address_list);
+free_ucp_addr:
+    ucs_free(ucp_addr);
+out:
+    return status;
+}
+
+static ssize_t ucp_cm_client_priv_pack_cb(void *arg, const char *dev_name,
+                                          void *priv_data)
+{
+    ucp_wireup_sockaddr_data_t *sa_data = priv_data;
+    ucp_ep_h ep                         = arg;
+    ucp_worker_h worker                 = ep->worker;
+    uct_cm_h cm                         = worker->cms[/*cm_idx = */ 0].cm;
+    uint64_t tl_bitmap;
+    ucp_wireup_ep_t *wireup_ep;
+    uct_ep_h tl_ep;
+    uct_cm_attr_t cm_attr;
+    uct_ep_params_t tl_ep_params;
+    void* ucp_addr;
+    size_t ucp_addr_size;
+    ucs_status_t status;
+    ucp_lane_index_t lane_idx;
+    ucp_rsc_index_t rsc_idx;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    wireup_ep = ucp_ep_get_cm_wireup_ep(ep);
+    ucs_assert_always(wireup_ep != NULL);
+    strncpy(wireup_ep->dev_name, dev_name, UCT_DEVICE_NAME_MAX);
+    wireup_ep->dev_name[UCT_DEVICE_NAME_MAX - 1] = '\0';
+
+    status = ucp_cm_preinit_client_lanes(ep);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    cm_attr.field_mask = UCT_CM_ATTR_FIELD_MAX_CONN_PRIV;
+    status             = uct_cm_query(cm, &cm_attr);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    tl_bitmap = 0;
+    for (lane_idx = 0; lane_idx < ucp_ep_num_lanes(ep); ++lane_idx) {
+        if (lane_idx == ucp_ep_get_cm_lane(ep)) {
+            continue;
+        }
+
+        rsc_idx = ucp_ep_get_rsc_index(ep, lane_idx);
+        if (rsc_idx == UCP_NULL_RESOURCE) {
+            continue;
+        }
+
+        tl_bitmap |= UCS_BIT(rsc_idx);
+        if (ucp_worker_iface_is_tl_p2p(ucp_ep_get_iface_attr(ep, lane_idx))) {
+            tl_ep_params.field_mask = UCT_EP_PARAM_FIELD_IFACE;
+            tl_ep_params.iface      = ucp_worker_iface(worker, rsc_idx)->iface;
+            status = uct_ep_create(&tl_ep_params, &tl_ep);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            ucp_wireup_assign_lane(ep, lane_idx, tl_ep, "sockaddr TL lane");
+        } else {
+            ucs_assert(ucp_ep_get_iface_attr(ep, lane_idx)->cap.flags &
+                       UCT_IFACE_FLAG_CONNECT_TO_IFACE);
+        }
+    }
+
+    /* Don't pack the device address to reduce address size, it will be
+     * delivered by uct_listener_conn_request_callback_t in
+     * uct_cm_remote_data_t */
+    status = ucp_address_pack(worker, ep, tl_bitmap,
```

Comment:
so cm_lane is not part of tl_bitmap? does not seem right..

Reply:
why? CM is not a TL

Reply:
i meant the actual transport which may be contained in it as next_ep

Reply:
text_ep on wireup CM lane is cm_lane, not a transport lane